res/res_pjsip_device_features.c

```diff
@@ -0,0 +1,1005 @@
+}
+
+#define DEVICE_FEATURE_UPDATE_STR(endpoint, db, upd, var, value, dbvalue) { \
+	ast_copy_string(feature_state_sub->sync_data.var, value, sizeof(feature_state_sub->sync_data.var)); \
+	feature_state_sub->sync_data.upd = 1; \
+	ast_db_put("subscription_device_features/" db, endpoint, dbvalue); \
+}
+
+static int parse_incoming_xml(struct feature_state_subscription *feature_state_sub, char *xmlbody, int xmllen, const char *endpoint_name, const char *number)
+{
+	int res = 0;
+	const char *nodename;
+	const char *tmpstr;
+	struct ast_xml_node *root, *tmpnode = NULL;
+	struct ast_xml_doc *xmldoc = ast_xml_read_memory(xmlbody, xmllen);
+
+	if (!xmldoc) {
+		ast_log(LOG_WARNING, "Failed to parse as XML: %.*s\n", xmllen, xmlbody);
+		return -1;
+	}
+
+	/*
+	 * Broadworks 11-BD5196-00 (2012)
+	 *
+	 * <?xml version="1.0" encoding="ISO-8859-1"?>
+	 * <SetDoNotDisturb xmlns="http://www.ecma-international.org/standards/ecma-323/csta/ed3">
+	 *    <device>7659366</device>
+	 *    <doNotDisturbOn>true</doNotDisturbOn>
+	 * </SetDoNotDisturb>
+	 *
+	 * (OR)
+	 *
+	 * <?xml version="1.0" encoding="ISO-8859-1"?>
+	 * <SetForwarding xmlns="http://www.ecma-international.org/standards/ecma-323/csta/ed3">
+	 *    <device>7659366</device>
+	 *    <activateForward>true</activateForward>
+	 *    <forwardingType>forwardImmediate</forwardingType>
+	 *    <forwardDN>2424</forwardDN>
+	 *    <ringCount></ringCount>
+	 * </SetForwarding>
+	 */
+
+	if (DEBUG_ATLEAST(1)) {
+		char *doc_str = NULL;
+		int doc_len = 0;
+		ast_xml_doc_dump_memory(xmldoc, &doc_str, &doc_len);
+		ast_debug(4, "Incoming doc len: %d\n%s\n", doc_len, doc_len ? doc_str : "<empty>");
+		ast_xml_free_text(doc_str);
+		doc_str = NULL;
+		doc_len = 0;
+	}
+
+	/* I can't figure out how to do this with the presence XML functions, which seem to be used for writing only at this point.
+	 * So we use the core XML routines. */
+	root = ast_xml_get_root(xmldoc);
+	if (!root) {
+		ast_log(LOG_WARNING, "No root?\n");
+		res = -1;
+		goto cleanup;
+	}
+	nodename = ast_xml_node_get_name(root);
+
+	/* The device is not used by Broadworks, and the phone can set this to any value. See 11-BD5196-00, 3.1.2 */
+	tmpnode = ast_xml_find_child_element(root, "device", NULL, NULL);
+	tmpstr = ast_xml_get_text(tmpnode);
+	ast_copy_string(feature_state_sub->sync_data.deviceid, tmpstr, sizeof(feature_state_sub->sync_data.deviceid)); /* Store the device ID so we can use it in the NOTIFY. */
+
+	if (!strcmp(nodename, "SetDoNotDisturb")) {
+		int dnd_disp;
+
+		tmpnode = ast_xml_find_child_element(root, "doNotDisturbOn", NULL, NULL);
+		tmpstr = ast_xml_get_text(tmpnode);
+		dnd_disp = ast_true(tmpstr) ? 1 : 0; /* ast_true returns -1 on success */
+		ast_xml_free_text(tmpstr);
+
+		if (!res) {
+			res = send_ami(feature_state_sub->sync_data.deviceid, endpoint_name, number, "DoNotDisturb", dnd_disp, 0, NULL);
+		}
+		if (!res) {
+			/* Note: We do NOT set sync_data.dnd = dnd_disp directly here! That's not our job!!!
```

Comment:
There needs to be a large comment at the top of this module detailing specifically how it works, the interactions, and the responsibilities of all entities involved.